### PR TITLE
Support `getParameter`'s parameters

### LINKED ISSUE #449: Support `getParameter`'s parameters
Currently, FerretDB behaves as if `getParameter` command was invoked as `{"getParameter": "*"}`. We should also support the other form `{"getParameter": <something>, "<parameter>": <something>}` that returns the value of the named parameter.

We should also test the edge cases where something is something weird like NaN, Infinity, etc.

See also #533.

### LINKED ISSUE #533: Support protocol errors without code
`mongosh` query `db.adminCommand({getParameter: "invalid"})` makes MongoDB return:

```
{
  "Checksum": 0,
  "FlagBits": 0,
  "Sections": [
    {
      "Document": {
        "$k": [
          "ok",
          "errmsg"
        ],
        "ok": {
          "$f": 0
        },
        "errmsg": "no option found to get"
      },
      "Kind": 0
    }
  ]
}
```

That error does not have an error code at all.

FerretDB code assumes that all errors has code [there](https://github.com/FerretDB/FerretDB/blob/ad8acc08a70f64ff4e877e594653a9146b738b0e/internal/handlers/common/error.go#L59-L61), [there](https://github.com/FerretDB/FerretDB/blob/ad8acc08a70f64ff4e877e594653a9146b738b0e/integration/query_array_test.go#L126-L130) and in other similar places.

We should make FerretDB compatible with MongoDB even in that weird case.

See also #449.

### PR DESCRIPTION
**Pull request by fenogentov**
Closes #449.
Closes #533.

### CODE REVIEW

**Review by seeforschauer — CHANGES_REQUESTED** [review]
@fenogentov check dirty is not clean: you need to run `task init` locally and see what changes.

**Comment by seeforschauer** [comment]
> > check dirty is not clean: you need to run `task init` locally and see what changes.
> 
> local test was ok

yes, it is ok.. and go generate -x ./... will add some changes which it's necessary to commit.

File: internal/handlers/common/error.go [hunk 1]
```diff
@@ -29,6 +29,8 @@ type ErrorCode int32
 
 const (
 	// For ProtocolError only.
+	ErrWeird = ErrorCode(0) // WeirdError
```
Comment (AlekSi) [inline internal/handlers/common/error.go:32]: It is weird for sure, but let's call it `ErrUnset` or `ErrUnknown`

**Review by AlekSi — CHANGES_REQUESTED** [review]
There are two more things we should do:

1. Add an integration test with `getParameter` call that shows the problem.
2. Do not set `code` and `codeName` for code == 0 there: https://github.com/FerretDB/FerretDB/blob/3b082cd08832682aee502f7b972742cbc35c71fb/internal/handlers/common/error.go#L98-L99

File: internal/handlers/common/error.go (hunk 1 above)
Comment (AlekSi) [reply internal/handlers/common/error.go:32]: Actually, `errUnset` – we don't have to export it

File: internal/handlers/common/error.go [hunk 2]
```diff
@@ -29,6 +29,8 @@ type ErrorCode int32
 
 const (
 	// For ProtocolError only.
```
Comment (AlekSi) [inline internal/handlers/common/error.go:31]: That comment is for `errInternalError`

File: internal/handlers/pg/msg_getparameter.go [hunk 3]
```diff
@@ -25,25 +26,83 @@ import (
...
+		"quiet", false,
+		"ok", float64(1),
+	)
 
 	var reply wire.OpMsg
+	var errNoFound error
+	var resDoc *types.Document
+	if getPrm == "*" {
+		resDoc = resDB
+
```
Comment (AlekSi) [inline internal/handlers/pg/msg_getparameter.go:52]: CI complains about that empty line. It should be removed by `task fmt`

File: integration/commands_administration_test.go [hunk 4]
```diff
@@ -175,21 +175,72 @@ func TestCommandsAdministrationCreateDropListDatabases(t *testing.T) {
 }
 
 func TestCommandsAdministrationGetParameter(t *testing.T) {
-	t.Parallel()
+	//	t.Parallel()
```
Comment (AlekSi) [inline integration/commands_administration_test.go:178]: Let's keep it

File: integration/query_array_test.go [hunk 5]
```diff
@@ -42,7 +42,7 @@ func TestQueryArraySize(t *testing.T) {
 	for name, tc := range map[string]struct {
 		filter      bson.D
 		expectedIDs []any
-		err         mongo.CommandError
+		err         *mongo.CommandError
```
Comment (AlekSi) [inline integration/query_array_test.go:45]: We merged more tests where we should do the same changes

File: integration/commands_administration_test.go [hunk 6]
```diff
@@ -175,21 +175,72 @@ func TestCommandsAdministrationCreateDropListDatabases(t *testing.T) {
...
+			err:    &mongo.CommandError{Code: 0, Name: "", Message: `no option found to get`},
+		},
+		"OnlyNonexistentParameters": {
+			filter: bson.D{{"getParameter", 1}, {"quiet_other", 1}},
+			err:    &mongo.CommandError{Code: 0, Name: "", Message: `no option found to get`},
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			//			t.Parallel()
```
Comment (AlekSi) [inline integration/commands_administration_test.go:224]: Let's run those tests in parallel

**Review by AlekSi — CHANGES_REQUESTED** [review]
I think it might be easier to make a separate PR from the current state of `main` branch that changes `err mongo.CommandError` to make it a pointer with matching `if tc.err.Code != 0` change, then return to that PR

File: integration/commands_administration_test.go [hunk 7]
```diff
@@ -185,18 +185,69 @@ func TestCommandsAdministrationGetParameter(t *testing.T) {
 		databaseName: "admin",
 	})
 
-	var actual bson.D
-	err := collection.Database().RunCommand(ctx, bson.D{{"getParameter", "*"}}).Decode(&actual)
-	require.NoError(t, err)
+	for name, tc := range map[string]struct {
+		filter     bson.D
+		expected   map[string]any
+		noExpected []string
```
Comment (AlekSi) [inline integration/commands_administration_test.go:191]:
```suggestion
		unexpected []string
```

File: internal/handlers/common/error.go [hunk 8]
```diff
@@ -28,6 +28,8 @@ import (
 type ErrorCode int32
 
 const (
+	errUnset = ErrorCode(0) //
```
Comment (AlekSi) [inline internal/handlers/common/error.go:31]:
```suggestion
	errUnset = ErrorCode(0) // Unset
```

And run `task gen` after that

**Review by AlekSi — CHANGES_REQUESTED** [review]
Please also do the second point from https://github.com/FerretDB/FerretDB/pull/535#pullrequestreview-953855776

File: internal/handlers/common/error.go (hunk 8 above)
Comment (seeforschauer) [reply internal/handlers/common/error.go:32]: @fenogentov why didn't you commit that change? for stringer comments matter

Comment (fenogentov) [reply internal/handlers/common/error.go]: > @fenogentov why didn't you commit that change? for stringer comments matter

mongo has this error with no name
*fixed for similar behavior

File: internal/handlers/common/error.go [hunk 9]
```diff
@@ -92,12 +92,15 @@ func (e *Error) Unwrap() error {
 
 // Document returns wire protocol error document.
 func (e *Error) Document() *types.Document {
-	return types.MustNewDocument(
+	d := types.MustNewDocument(
```
Comment (seeforschauer) [inline internal/handlers/common/error.go:95]: `	d := must.NotFail(types.NewDocument(`

File: internal/handlers/pg/msg_getparameter.go [hunk 10]
```diff
@@ -25,25 +26,82 @@ import (
...
+		"acceptApiVersion2", false,
+		"authSchemaVersion", int32(5),
+		"quiet", false,
+		"ok", float64(1),
+	)
 
 	var reply wire.OpMsg
+	var errNoFound error
+	var resDoc *types.Document
+	if getPrm == "*" {
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_getparameter.go:50]: let's create fully meaningful variables and name it `getParam` for example

File: internal/handlers/pg/msg_getparameter.go [hunk 11]
```diff
@@ -25,25 +26,82 @@ import (
...
+			return nil, nil, err
+		}
+	}
+
+	if doc.Len() < 1 {
+		err := doc.Set("ok", float64(0))
+		if err != nil {
+			return nil, nil, err
+		}
+		errNoFound = common.NewErrorMsg(common.ErrorCode(0), "no option found to get")
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_getparameter.go:98]: here you can make an early return and the other part of code will become more readable

File: internal/handlers/pg/msg_getparameter.go (hunk 10 above)
Comment (fenogentov) [reply internal/handlers/pg/msg_getparameter.go:50]: ok

File: internal/handlers/pg/msg_getparameter.go (hunk 11 above)
Comment (fenogentov) [reply internal/handlers/pg/msg_getparameter.go:98]: fix

File: internal/handlers/common/error.go (hunk 9 above)
Comment (fenogentov) [reply internal/handlers/common/error.go:95]: overlooked. fix

File: internal/handlers/pg/msg_getparameter.go [hunk 12]
```diff
@@ -25,25 +26,82 @@ import (
...
+	)
 
 	var reply wire.OpMsg
+	var errNoFound error
+	var resDoc *types.Document
+	if getParameter == "*" {
+		resDoc = resDB
+	} else {
+		resDoc, errNoFound, err = selectParam(cmd, resDB)
+		if err != nil {
```
Comment (seeforschauer) [inline internal/handlers/pg/msg_getparameter.go:54]: I would suggest here:
```
	resDoc = resDB
	if getParameter != "*" {
	  resDoc, errNoFound, err = selectParam(cmd, resDB)
	  if err != nil {
```

File: integration/commands_administration_test.go [hunk 13]
```diff
@@ -185,18 +185,69 @@ func TestCommandsAdministrationGetParameter(t *testing.T) {
...
+			},
+			unexpected: []string{"quiet_other"},
+		},
+		"EmptyParameters": {
+			command: bson.D{{"getParameter", 1}},
+			err:     &mongo.CommandError{Code: 0, Name: "", Message: `no option found to get`},
+		},
+		"OnlyNonexistentParameters": {
+			command: bson.D{{"getParameter", 1}, {"quiet_other", 1}},
+			err:     &mongo.CommandError{Code: 0, Name: "", Message: `no option found to get`},
```
Comment (AlekSi) [inline integration/commands_administration_test.go:224]:
```suggestion
			err:     &mongo.CommandError{Message: `no option found to get`},
```

Zero values are good enough

File: integration/commands_administration_test.go [hunk 14]
```diff
@@ -185,18 +185,69 @@ func TestCommandsAdministrationGetParameter(t *testing.T) {
...
-	err := collection.Database().RunCommand(ctx, bson.D{{"getParameter", "*"}}).Decode(&actual)
-	require.NoError(t, err)
+	for name, tc := range map[string]struct {
+		command    bson.D
+		expected   map[string]any
+		unexpected []string
+		err        *mongo.CommandError
+	}{
+		"AllParameters": {
+			command: bson.D{{"getParameter", "*"}},
```
Comment (AlekSi) [inline integration/commands_administration_test.go:195]: Let's add a test for `bson.D{{"getParameter", "*"}, {"quiet", 1}, {"quiet_other", 1}}`

File: integration/commands_administration_test.go [hunk 15]
```diff
@@ -185,18 +185,69 @@ func TestCommandsAdministrationGetParameter(t *testing.T) {
...
+			expected: map[string]any{
+				"acceptApiVersion2": false,
+				"authSchemaVersion": int32(5),
+				"quiet":             false,
+				"ok":                float64(1),
+			},
+		},
+		"ExistingParameters": {
+			command: bson.D{{"getParameter", 1}, {"quiet", 1}},
```
Comment (AlekSi) [inline integration/commands_administration_test.go:204]: What would happen for `{"getParameter", 0}`? `{"getParameter", NULL}`? Should we add tests for those cases?

File: internal/handlers/pg/msg_getparameter.go [hunk 16]
```diff
@@ -25,25 +26,80 @@ import (
 
 // MsgGetParameter OpMsg used to get parameter.
 func (h *Handler) MsgGetParameter(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, error) {
-	document, err := msg.Document()
+	cmd, err := msg.Document()
```
Comment (AlekSi) [inline internal/handlers/pg/msg_getparameter.go:29]: All handlers has `document, err := msg.Document()` as the first line. Let's keep it the same for consistency

File: internal/handlers/pg/msg_getparameter.go [hunk 17]
```diff
@@ -25,25 +26,80 @@ import (
...
 	if err != nil {
 		return nil, lazyerrors.Error(err)
 	}
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/449
+	getParameter, err := cmd.Get("getParameter")
```
Comment (AlekSi) [inline internal/handlers/pg/msg_getparameter.go:34]: That's typically written as `command := document.Command()`

File: internal/handlers/pg/msg_getparameter.go [hunk 18]
```diff
@@ -25,25 +26,80 @@ import (
...
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
 
-	common.Ignored(document, h.l, "comment")
+	// SELECT * FROM 'admin'
```
Comment (AlekSi) [inline internal/handlers/pg/msg_getparameter.go:39]: What does this mean?

File: internal/handlers/pg/msg_getparameter.go [hunk 19]
```diff
@@ -25,25 +26,80 @@ import (
...
 		return nil, lazyerrors.Error(err)
 	}
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/449
+	getParameter, err := cmd.Get("getParameter")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
 
-	common.Ignored(document, h.l, "comment")
```
Comment (AlekSi) [inline internal/handlers/pg/msg_getparameter.go:35]: We should keep that line

File: internal/handlers/pg/msg_getparameter.go [hunk 20]
```diff
@@ -25,25 +26,80 @@ import (
...
-	return &reply, nil
+	comment, err := cmd.Get("comment")
+	if err == nil {
+		common.Ignored(cmd, h.l, fmt.Sprint(comment))
+	}
+
+	return &reply, errNoFound
+}
+
+// selectParam is makes a selection of the requested parameters received from database 'admin'.
```
Comment (AlekSi) [inline internal/handlers/pg/msg_getparameter.go:72]: Well, that's not what function does. We don't have that database

File: internal/handlers/pg/msg_getparameter.go [hunk 21]
```diff
@@ -25,25 +26,80 @@ import (
...
+func selectParam(command, resDB *types.Document) (doc *types.Document, errNoFound error, err error) {
```
Comment (AlekSi) [inline internal/handlers/pg/msg_getparameter.go:73]: TBH, it is hard to tell what that function does with parameters, two returned errors, etc.

After all, all we need to do is to get values from fixed document: either one or all of them. It should not be that complicated

File: internal/handlers/pg/msg_getparameter.go [hunk 22]
```diff
@@ -25,25 +26,80 @@ import (
...
+		Documents: []*types.Document{resDoc},
 	})
 	if err != nil {
 		return nil, lazyerrors.Error(err)
 	}
 
-	return &reply, nil
+	comment, err := cmd.Get("comment")
+	if err == nil {
+		common.Ignored(cmd, h.l, fmt.Sprint(comment))
```
Comment (AlekSi) [inline internal/handlers/pg/msg_getparameter.go:66]: There is not reason to make it that complicated, with `if`, error handling, etc. The old line of code is good

**Review by AlekSi — CHANGES_REQUESTED** [review]

File: integration/commands_administration_test.go (hunk 13 above)
Comment (fenogentov) [reply integration/commands_administration_test.go:224]: fix

File: integration/commands_administration_test.go (hunk 14 above)
Comment (fenogentov) [reply integration/commands_administration_test.go:195]: fix

File: integration/commands_administration_test.go (hunk 15 above)
Comment (fenogentov) [reply integration/commands_administration_test.go:204]: fix

File: internal/handlers/pg/msg_getparameter.go (hunk 16 above)
Comment (fenogentov) [reply internal/handlers/pg/msg_getparameter.go:29]: fix

File: internal/handlers/pg/msg_getparameter.go (hunk 22 above)
Comment (fenogentov) [reply internal/handlers/pg/msg_getparameter.go:66]: that is, we ignore the comment from command?

File: internal/handlers/pg/msg_getparameter.go (hunk 20 above)
Comment (fenogentov) [reply internal/handlers/pg/msg_getparameter.go:72]: fix

**Review by seeforschauer — CHANGES_REQUESTED** [review]
@fenogentov 
would you be so kind to set the PR status to Draft if the work is in progress? 
Draft status can be changed under the reviewer's list.
Thank you.

File: internal/handlers/pg/msg_getparameter.go (hunk 17 above)
Comment (fenogentov) [reply internal/handlers/pg/msg_getparameter.go]: `command := document.Command()` this is getting the command name
and in this case I get the value of the command

**Review by seeforschauer — CHANGES_REQUESTED** [review]
LGTM, could you please resolve conflicts?

File: integration/commands_administration_test.go (hunk 13 above)
Comment (seeforschauer) [reply integration/commands_administration_test.go:273]: hmmm, looks like there are some more similar lines of code, could you please fix them also?

**Review by seeforschauer — CHANGES_REQUESTED** [review]

File: integration/commands_administration_test.go (hunk 13 above)
Comment (fenogentov) [reply integration/commands_administration_test.go:273]: all tests are different.
can't find the same

Comment (seeforschauer) [reply integration/commands_administration_test.go:273]: I mean, you wrote that it's fixed and it looks like you didn't push.  It is actually still in the code.

**Review by seeforschauer — DISMISSED** [review]
Thank you, @fenogentov!

File: internal/handlers/pg/msg_getparameter.go (hunk 17 above)
Comment (AlekSi) [reply internal/handlers/pg/msg_getparameter.go]: That's true, but then we should use `document.Get(command)`, not `command.Get("getParameter")`, for consistencty

**Review by AlekSi — CHANGES_REQUESTED** [review]
Please remove `integration/integration.txt` from the PR, it should not be commited

File: internal/handlers/pg/msg_getparameter.go (hunk 22 above)
Comment (AlekSi) [reply internal/handlers/pg/msg_getparameter.go]: We ignore comment either way, aren't we? That's what `common.Ignored` does

**Comment by seeforschauer** [comment]
@fenogentov would you please resolve conflicts?

**Review by AlekSi — APPROVED** [review]

**Pull request merged**